### ISSUE TITLE
ddtrace 1.19.x to 2.1.x prevent app from exiting cleanly

### ISSUE THREAD

**Issue by lawrenceong** [DataDog/dd-trace-py#7273]
### Summary of problem

When trying to exit an app by sending a TERM signal to it, ddtrace is preventing the app from exiting

### Which version of dd-trace-py are you using?

`<= 1.18.x` - not an issue
`>= 1.19.x to 2.1.x` - broken

### Which version of pip are you using?

```
pip 23.2.1 from /.../3.11.5/lib/python3.11/site-packages/pip (python 3.11)
```

### Which libraries and their versions are you using?

```
ddtrace = "^2.1.5"
flask = "^3.0.0"
```

<details>
<summary>pip freeze</summary>
attrs==23.1.0
blinker==1.6.3
bytecode==0.15.1
cattrs==23.1.2
click==8.1.7
ddsketch==2.0.4
ddtrace==2.1.5
Deprecated==1.2.14
envier==0.4.0
Flask==3.0.0
importlib-metadata==6.8.0
itsdangerous==2.1.2
Jinja2==3.1.2
MarkupSafe==2.1.3
opentelemetry-api==1.20.0
protobuf==4.24.4
six==1.16.0
typing_extensions==4.8.0
Werkzeug==3.0.1
wrapt==1.15.0
xmltodict==0.13.0
zipp==3.17.0
</details>

### How can we reproduce your problem?
Create a simple app that looks like the following:

`app.py`
```
from flask import Flask
app = Flask(__name__)


@app.route('/')
def hello_world():
    return 'Hello, World!'


if __name__ == '__main__':
    app.run(port=8082)
```

Create a simple runner that looks like the following:

`run.sh`
```
ddtrace-run python ./app.py
```

Run the app.  It will come up looking like this:

```
 * Serving Flask app 'app'
 * Debug mode: off
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:8082
Press CTRL+C to quit
```

Find the pid of the running app:

```
ps -ef | grep app.py$
  501 99043 98370   0  1:59pm ttys004    0:00.36 //virtualenvs/hangonexit-vrlmzGGs-py3.11/bin/python ./app.py
```

Send a SIGTERM to the app:

```
kill 99043
```

### What is the result that you get?
App does not exit when killed the first time.  App will exit if killed the second time.  It will output the following stack trace:

```
Traceback (most recent call last):
  File "//ddtrace/hangonexit/./runme.py", line 11, in <module>
    app.run(host="0.0.0.0", port=8082)
  File "//virtualenvs/hangonexit-vrlmzGGs-py3.11/lib/python3.11/site-packages/flask/app.py", line 612, in run
    run_simple(t.cast(str, host), port, self, **options)
  File "//virtualenvs/hangonexit-vrlmzGGs-py3.11/lib/python3.11/site-packages/werkzeug/serving.py", line 1109, in run_simple
    srv.serve_forever()
  File "//virtualenvs/hangonexit-vrlmzGGs-py3.11/lib/python3.11/site-packages/werkzeug/serving.py", line 806, in serve_forever
    super().serve_forever(poll_interval=poll_interval)
  File "//.pyenv/versions/3.11.5/lib/python3.11/socketserver.py", line 233, in serve_forever
    ready = selector.select(poll_interval)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "//.pyenv/versions/3.11.5/lib/python3.11/selectors.py", line 415, in select
    fd_event_list = self._selector.poll(timeout)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "//virtualenvs/hangonexit-vrlmzGGs-py3.11/lib/python3.11/site-packages/ddtrace/internal/atexit.py", line 68, in handle_exit
    f()
  File "//virtualenvs/hangonexit-vrlmzGGs-py3.11/lib/python3.11/site-packages/ddtrace/tracer.py", line 292, in _atexit
    self.shutdown(timeout=self.SHUTDOWN_TIMEOUT)
  File "//virtualenvs/hangonexit-vrlmzGGs-py3.11/lib/python3.11/site-packages/ddtrace/tracer.py", line 1004, in shutdown
    forksafe.unregister(self._child_after_fork)
  File "//virtualenvs/hangonexit-vrlmzGGs-py3.11/lib/python3.11/site-packages/ddtrace/internal/forksafe.py", line 56, in unregister
    _registry.remove(after_in_child)
ValueError: list.remove(x): x not in list
```

**NOTE**: This does not happen with ddtrace version <= 1.18.x.  App exits straight away.

### What is the result that you expected?
App will exit cleanly when killed using SIGTERM

**Comment by lawrenceong** [comment]
@romainkomorndatadog @majorgreys Any update on the investigation?  We've had to stop the library upgrade due to the bug and the older version does not work with certain newer plugins, eg. openai.

**Comment by wrkhenddher** [comment]
FWIW, in our case, ^C'ing after hanging yields the following (not sure if it's the same issue though it's similar to my colleague's https://github.com/DataDog/dd-trace-py/issues/6894):

ddtrace 1.20.10

```
^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/Users/henddher/.pyenv/versions/3.8.16/lib/python3.8/multiprocessing/popen_fork.py", line 27, in poll
    pid, sts = os.waitpid(self.pid, flag)
  File "/Users/henddher/Documents/github/gdcapi_gene_expr/venv/lib/python3.8/site-packages/ddtrace/internal/utils/signals.py", line 16, in wrap_signals
    old_signal(*args, **kwargs)
KeyboardInterrupt
```

**Comment by lawrenceong** [comment]
Confirmed fixed in 1.9.7, 1.20.13, 2.0.10, 2.1.8.  Still broken in 2.2.0.
